Ena 4969

haseeb-gh opened the pull request.

haseeb-gh committed:
ENA-4969

- removed code that slowed down validation.
- replaced some code with optimized one that should lower validation time.
- refactored and added comments to some code to make it more descriptive and easier to understand.
- discovered and fixed is issue that was not returning reads with normalized quality scores. added a unit tests for this as well.

haseeb-gh committed:
ENA-4969

- overhauled the fastq validation implementation.
- made changes necessary to support more than 2 paired fastqs.
- refactored, simplified and optimized code.
- fixed some potential bugs.
- removed obselete, updated exsiting and added new tests.
- added lots of comments to well document things.

haseeb-gh committed:
ENA-4969

- updated comments and one unit test.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/FileCompression.java:
```diff
@@ -50,14 +50,14 @@
             {
             case BZ2:
-                is = new BZip2CompressorInputStream( new BufferedInputStream( is ), true );
+                is = new BZip2CompressorInputStream( new BufferedInputStream( is, (1024 * 1024) ), true );
```
MIght want to create a static final constant for 1024 * 1024.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/FastqReadReadConverter.java:
```diff
@@ -10,18 +10,16 @@
- * Detects quality normalizer automatically based on the given file.
+ * Similar to {@link ReadConverter}. except here, quality normalization strategy is chosen automatically.
  */
 public class FastqReadReadConverter extends AbstractReadConverter<Read> {
```
- Read type could be created from any source that has name, bases, quality scores. This is commonly called  'Fastq level' data.
- Suggestion to call this class: NormalizeQualityReadConverter.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/ReadConverter.java:
```diff
@@ -10,22 +10,44 @@
+ * Similar to {@link FastqReadReadConverter}, but here, base quality normalizer is provided explicitly. If no normalizer
+ * is provided then normalization is not performed.
+ */
 public class ReadConverter extends AbstractReadConverter<Read> {
```
- Read type could be created from any source that has name, bases, quality scores. This is commonly called  'Fastq level' data.
- Suggestion to call this class: PreserveQualityReadConverter.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/ReadConverter.java:
```diff
@@ -10,22 +10,44 @@
 
     private final ReadReader readReader;
 
+    public ReadConverter(InputStream istream, String defaultAttr) {
```
Just a stylistic comment. When you start having a lot of constructors it may be better to use the builder pattern e.g.:

```
private ReadConverter(InputStream istream, Long readLimit, QualityNormalizer normalizer, String defaultAttr) {
...
}

public static ReadConverterBuilder builder(InputStream istream,  String defaultAttr) {
   return new ReadConverterBuilder(istream, defaultAttr);
}


public static class ReadConverterBuilder {
   private InputStream istream;
   ...
   public ReadConverterBuilder(InputStream istream,  String defaultAttr) {
      ...
   }

   public ReadConverterBuilder readLimit(long readLimit) {
      ...
      return this;
   }
   ...
   public ReadConverter build() {
      return new ReadConverter(...
   }
}
```

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/ReadReader.java:
```diff
@@ -117,24 +118,42 @@ public DataSpotReaderParams(Long      line_no,
     private final String defaultReadIndex;
 
-    private final QualityNormalizer normalizer;
+    private final Consumer<Read> readNormalizationConsumer;
```
Suggestion to rename to qualityNormalizer.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/ReadReader.java:
```diff
@@ -117,24 +118,42 @@ public DataSpotReaderParams(Long      line_no,
+    /**
+     * Creates a reader that does not perform any quality normalization.
+     */
+    public ReadReader(String readIndex) {
```
We have quite a few constructors and might be better to use a builder pattern e.g. both the quality normalizer and expectedLength are optional parameters.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/FastqIterativeWriter.java:
```diff
@@ -31,6 +31,8 @@
     private READ_TYPE read_type;
     private QualityNormalizer[] normalizers;
```
- I would remove all setters and getters from FastqIterativeWriter as they are not used or needed.
-  would make files and read type private finals and require them to be provided in the constructor.
- normalizers would not be needed if we add
```
private final BiFunction<File, int, ReadConverter> readConverterFactory readConverterFactory
```

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/FastqIterativeWriter.java:
```diff
@@ -10,11 +10,11 @@
-import uk.ac.ebi.ena.readtools.common.reads.QualityNormalizer;
-
 public class
 FastqIterativeWriter implements Iterable<PairedRead> {
```
I think FastqIterativeWriter should accept
```
BiFunction<File, int, ReadConverter> readConverterFactory
```

as a parameter rather than readLimit and normalizer param values needed to construct the ReadConverter.

i.e. I think better would be able to do this when creating the FastqIterativeWriter: (private final and add to the constructor):

```
(file, readIndex) -> new ReadConverter(FileCompression.open(file), readLimit, normalizer, readIndex);
```

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/ReadConverter.java:
```diff
@@ -10,22 +10,44 @@
 
     private final ReadReader readReader;
 
+    public ReadConverter(InputStream istream, String defaultAttr) {
```
The defaultAttr name is confusing and we should use either readIndex or defaultReadIndex depending on the context.

haseeb-gh replied:
I tried but im afraid it is not very straight forward to do that. With this field being String, there are many places in code where an empty string "" is being set for this paramater. To replace it with int or even Integer, i will need to look at every where this field is being parsed into integer and understand how parsing code handles null/empty value and how surrounding code is affected by all of that. For now, I will just rename the field defaultAttr or default_attr to defaultReadIndex to remove name confusion. Maybe in future we can take a detailed look the idea of a default read index is implemented and how to go about changing it.

raskoleinonen replied:
Thank you, renaming the field will help a lot :-)

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/FastqReadReadConverter.java:
```diff
@@ -41,12 +39,12 @@ public FastqReadReadConverter(InputStream istream, String defaultAttr, String fi
      */
-    public FastqReadReadConverter(InputStream istream, Duration runDuration, String defaultAttr, String filePath) {
-        super(istream, runDuration);
+    public FastqReadReadConverter(InputStream istream, Long readLimit, String defaultAttr, String filePath) {
```
We should replace defaultAttr everywhere with either readIndex or defaultReadIndex (depending if it is immutable or mutable when reading > 1 file). Also maybe we could change this to an int as we are casting it to/from int in several places.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/fastq/FastqIterativeWriterIterator.java:
```diff
@@ -65,15 +69,28 @@
 
         for (File file : files) {
             final String default_attr = Integer.toString(attr++);
```
Should be called readIndex

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/validator/ReadsValidator.java:
```diff
@@ -139,21 +139,17 @@ public class ReadsValidator
-            FastqScanner fs = new FastqScanner(
-                    manifest.isQuick() ? DEFAULT_QUICK_RUN_DURATION : null, manifest.getPairingHorizon() ) {
+            //Higher expected size results in better accuracy of validation techniques used by FastqScanner.
+            int expectedSize = readLimit * 5;
```
expectedSize is bloom filter specific configuration parameter. Optimally it would not be visible to clients but would be set based from the readLimit when we create the bloomfilter internally.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/BloomWrapper.java:
```diff
@@ -19,59 +22,65 @@
     
     public 
-    BloomWrapper( long expected_reads )
+    BloomWrapper( long expectedReads )
```
Possibly this could take the readLimit as parameter and assign the size from this.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/BloomWrapper.java:
```diff
@@ -19,59 +22,65 @@
+    BloomWrapper( long expectedReads )
     {
-        this( expected_reads, 100_000 );
+        this( expectedReads, 100_000 );
```
I would make possibleDuplicatesRetainLimit 10 or something small like this.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/BloomWrapper.java:
```diff
@@ -19,59 +22,65 @@
-    private final int collect_max; 
-    private final Set<String> suspected;
+    private BloomFilter<String> bloom;
+    private double falsePositiveProbability = 0.01;
```
this could be static final -> falsePositiveProbability

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/FastqScanner.java:
```diff
@@ -75,9 +74,13 @@
+     * @param readLimit Only process limited number of reads per file. Prevents from processing all reads in a given file.
+     * @param expected_size
+     */
+    public FastqScanner(Long readLimit, int expected_size )
```
Suggested not to take expected_size as parameter.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/AbstractReadConverter.java:
```diff
@@ -10,25 +10,23 @@
-
 public abstract class
 AbstractReadConverter<T extends Spot> extends Thread implements Converter<T> {
     private static final int YIELD_CYCLES = 362;//16384;
```
ERROR_CHECK_BATCH_SIZE

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/loader/common/converter/AbstractReadConverter.java:
```diff
@@ -87,14 +85,13 @@ public final void run() {
             do {
                 synchronized (readWriter) {
-                    for (int yield = YIELD_CYCLES; yield > 0; --yield)
+                    for (int yield = YIELD_CYCLES; yield > 0 && keepRunning.get(); --yield)
```
for (int readCount = ERROR_CHECK_BATCH_SIZE; readCount > 0 ...

keepRunning  -> isWithinReadLimit

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/FastqScanner.java:
```diff
@@ -194,125 +199,175 @@ public boolean isOk() {
 
             Set<String> flabelset = new HashSet<>();
-            checkSingleFile( fileResult, rf, flabelset, pairing, duplications );
+            checkSingleFile( fileResult, rf, flabelset, mainFileAndCurrentFilePairing, currentFileDuplications );
```
prepareBloomWrappers

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/FastqScanner.java:
```diff
@@ -194,125 +199,175 @@ public boolean isOk() {
+                        e.getKey(),
+                        e.getValue().toString()))));
+
+                if( duplicationResult.isValid() ) {
```
No read duplicates found.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/FastqScanner.java:
```diff
@@ -194,125 +199,175 @@ public boolean isOk() {
+         * within the file only.
+         */
+
+        RawReadsFile mainFile = null;
```
Think if it is possible to move the mainFile stuff to happen before the loop.

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/FastqScanner.java:
```diff
@@ -194,125 +199,175 @@ public boolean isOk() {
+                mainFileDuplications = currentFileDuplications;
+                mainFilePairing = mainFileAndCurrentFilePairing;
+            } else {
+                long readCount = Math.max(mainFileDuplications.getAddCount(), currentFileDuplications.getAddCount());
```
This block could be extracted as a method:
```
addPairingPercentage()
```

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/FastqScanner.java:
```diff
@@ -194,125 +199,175 @@ public boolean isOk() {
+                        String.format( "Detected paired fastq submission with less than %d%% of paired reads between %s and %s",
+                            PAIRING_THRESHOLD, lowestPairingPercentagePair.fileName1, lowestPairingPercentagePair.fileName2 )));
+                } else {
+                    result.add(ValidationMessage.info("Pairing percentage is at acceptable level."));
```
I don't think we need to store this message -> ... acceptable ...

raskoleinonen commented on src/main/java/uk/ac/ebi/ena/readtools/webin/cli/rawreads/FastqScanner.java:
```diff
@@ -194,125 +199,175 @@ public boolean isOk() {
-            log.info( msg );
 
-            long index = 1;
+        String msg = "Performing additional checks for file " + rf.getFilename();
```
Checking for duplicate reads for file:

raskoleinonen commented on src/test/java/uk/ac/ebi/ena/readtools/fastq/ena/Fastq2SamTest.java:
```diff
@@ -25,22 +36,12 @@
+import static junit.framework.TestCase.fail;
 
+/**
+ * TODO The tests in here need to compare data between source fastq(s) and generated bam file as well.
```
BAM conversion should not be attempted except for 1 or 2 application read Fastq. We should not fail in other cases. We should just not try to create the bam.

haseeb-gh committed:
ENA-4969

- Simplified fastq scanner for better readibility.
- Refactored code. Added more comments and TODOs for important things to change in future.
- Updated tests.
